Capture CSP violations and report to Sentry.

glogiotatidis opened the pull request.

glogiotatidis commented:
@pmac r?

pmac commented on careers/base/views.py:
```diff
@@ -1,7 +1,41 @@
+        return HttpResponseBadRequest('Invalid CSP Report')
+
+    try:
+        blocked_uri = csp_data['csp-report']['blocked-uri']
```
Should we capture more info? Looking at the spec, we might also want to at least know the "document-uri" and possibly the "violated-directive".

pmac commented on careers/base/views.py:
```diff
@@ -1,7 +1,41 @@
+
+    client.captureMessage(
+        message='CSP Violation: {}'.format(blocked_uri),
+        data=data)
```
Oh. Nevermind. You're sending all of the data. Perfect! 💯

glogiotatidis replied:
Indeed, here's an example capture https://sentry.prod.mozaws.net/marketing/careers-stage-us-west/issues/348979/

pmac commented on careers/base/views.py:
```diff
@@ -1,7 +1,41 @@
+def csp_violation_capture(request):
+    data = client.get_data_from_request(request)
+    data.update({
+        'level': logging.INFO,
```
What's the point of this? Is this Sentry specific?

glogiotatidis replied:
Yes, it's used for event categorization in Sentry

pmac commented:
This looks fantastic. I can't wait to get this into all of our projects. Let me know how it goes. 

r+

pmac commented on careers/base/views.py:
```diff
@@ -1,7 +1,41 @@
+@csrf_exempt
+@require_POST
+def csp_violation_capture(request):
+    data = client.get_data_from_request(request)
```
Can't find any docs for this method. Where'd you find it?

glogiotatidis replied:
Indeed no documentation, I find it by reading `raven`'s code

pmac commented on careers/settings.py:
```diff
@@ -184,6 +187,7 @@
     'http://*.mozilla.net',
     'https://*.mozilla.net',
 )
+CSP_REPORT_URI = lazy(lambda: reverse('csp-violation-capture'), unicode)()
```
I'd add a `CSP_REPORT_ONLY` here as default `False` and configurable. I'm implementing this in my csp branch for bedrock now, and I did this:

```python
CSP_REPORT_ONLY = config('CSP_REPORT_ONLY', default=False, cast=bool)
```

It will allow you to quickly disable CSP on the site while still getting the reports so you can diagnose and fix.

pmac commented:
I pinged you on a couple of things in the bedrock PR. Mainly I got a little worried that we could DOS ourselves if we made a CSP mistake on a popular page since every violation generates a POST to the report URL. I added a setting that disables the report-uri header as well as makes the report view a very fast noop. I welcome your thoughts on the matter especially if there's a better strategy for mitigating this flood issue. We could always standup a service specifically for accepting CSP violation posts from our sites, but that seems like overkill at this point.

pmac commented:
I also just upgraded to django-csp==3.0 which supports the child-src directive (frame-src is deprecated) and it didn't work with the lazy URL for the reports. I switched to just using `/csp-violation-capture` as the default, which I think is fine, but worth noting.

glogiotatidis commented:
> I pinged you on a couple of things in the bedrock PR. Mainly I got a little worried that we could DOS ourselves if we made a CSP mistake on a popular page since every violation generates a POST to the report URL. I added a setting that disables the report-uri header as well as makes the report view a very fast noop. I welcome your thoughts on the matter especially if there's a better strategy for mitigating this flood issue. We could always standup a service specifically for accepting CSP violation posts from our sites, but that seems like overkill at this point.

Valid points. I also added the CSP_REPORT_ENABLE config

glogiotatidis commented:
> I also just upgraded to django-csp==3.0 which supports the child-src directive (frame-src is deprecated) and it didn't work with the lazy URL for the reports. I switched to just using /csp-violation-capture as the default, which I think is fine, but worth noting.

👍 Also upgraded

glogiotatidis commented:
Thanks!
